From 227e25920f9b0bd86f83c76857487f75e412ae60 Mon Sep 17 00:00:00 2001
From: Fang Hui <b31070@freescale.com>
Date: Wed, 16 Jul 2014 16:40:46 +0800
Subject: [PATCH] MA-6398 [Android_MX6Q_SD] Camera:Preview hung after taking a
 picture. 2%

in ION_IOC_PHYS, should call ion_handle_put after ion_handle_get_by_id,
or the ref count will added by 1. So when ion free, will not really free.

Signed-off-by: Fang Hui <b31070@freescale.com>
Signed-off-by: Liu Xiaowen <xiaowen.liu@freescale.com>
---
 drivers/staging/android/ion/ion.c         | 7 +++++++
 drivers/staging/android/ion/mxc/mxc_ion.c | 3 +++
 2 files changed, 10 insertions(+)

diff --git a/drivers/staging/android/ion/ion.c b/drivers/staging/android/ion/ion.c
index 136af9e..218a02c 100644
--- a/drivers/staging/android/ion/ion.c
+++ b/drivers/staging/android/ion/ion.c
@@ -3,6 +3,7 @@
  * drivers/staging/android/ion/ion.c
  *
  * Copyright (C) 2011 Google, Inc.
+ * Copyright (C) 2014-2015 Freescale Semiconductor, Inc.
  *
  * This software is licensed under the terms of the GNU General Public
  * License version 2, as published by the Free Software Foundation, and
@@ -400,6 +401,12 @@ static int ion_handle_put(struct ion_handle *handle)
 	return ret;
 }
 
+/* for used in mxc_ion.c */
+int ion_handle_put_wrap(struct ion_handle *handle)
+{
+	return ion_handle_put(handle);
+}
+
 static struct ion_handle *ion_handle_lookup(struct ion_client *client,
 					    struct ion_buffer *buffer)
 {
diff --git a/drivers/staging/android/ion/mxc/mxc_ion.c b/drivers/staging/android/ion/mxc/mxc_ion.c
index 204ee44..29b6d90 100644
--- a/drivers/staging/android/ion/mxc/mxc_ion.c
+++ b/drivers/staging/android/ion/mxc/mxc_ion.c
@@ -31,6 +31,7 @@ static int num_heaps = 1;
 static struct ion_heap **heaps;
 static int cacheable;
 
+extern int ion_handle_put_wrap(struct ion_handle *handle);
 extern struct ion_handle *ion_handle_get_by_id(struct ion_client *client,
 						 int id);
 
@@ -93,6 +94,8 @@ static long mxc_custom_ioctl(struct ion_client *client,
 			return PTR_ERR(handle);
 
 		valid = ion_phys(client, handle, &phys, &len);
+		ion_handle_put_wrap(handle);
+
 		if (valid)
 			return -1;
 		data.phys = phys;
-- 
1.8.0

